### ISSUE TITLE
FP16 inference produces black images (NaN latents)

### ISSUE THREAD

**Issue by Mitrandin** [Tongyi-MAI/Z-Image#14]
First of all, thank you for the great model! 

Issue:
Running the model in the torch.float16 either in comfyui or diffusers results in pure black image. float32 works as intended but very slow on my PC. bfloat16 is upcast to float32 internally.

System specs:
OS: Windows 10
GPU: NVIDIA RTX 2060 (6 GB VRAM)
RAM: 16 GB
PyTorch: 2.9.1+cu130
Diffusers: 0.36
ComfyUI: latest version
NO triton or sage-attention.

Steps to reproduce:
1) Run the basic comfyui workflow with startup flag --fp16-unet. 
Checked both quantized gguf model https://huggingface.co/jayn7/Z-Image-Turbo-GGUF/blob/main/z_image_turbo-Q4_K_M.gguf and https://huggingface.co/T5B/Z-Image-Turbo-FP8/blob/main/z-image-turbo-fp8-e4m3fn.safetensors.
Both of them produces black images (latent contains only nan values). 

2) Run diffusers workflow with https://huggingface.co/Disty0/Z-Image-Turbo-SDNQ-uint4-svd-r32 with provided code:
import torch
import diffusers
from sdnq import SDNQConfig # import sdnq to register it into diffusers and transformers

pipe = diffusers.ZImagePipeline.from_pretrained("Disty0/Z-Image-Turbo-SDNQ-uint4-svd-r32", torch_dtype=torch.float16)
pipe.enable_model_cpu_offload()

prompt = "any prompt"
image = pipe(
    prompt=prompt,
    height=1024,
    width=1024,
    num_inference_steps=9,
    guidance_scale=0.0,
    generator=torch.manual_seed(42),
).images[0]

→ Same result: black image. Inference in float32 works correctly, but is ~3× slower in ComfyUI and ~15× slower in Diffusers (no VRAM overflow—only partial CPU offloading occurs).

Notably:
Forcing torch.bfloat16 uses more VRAM than float32, yet inference speed remains the same as FP32 in Diffusers.
torch.cuda.is_bf16_supported() returns True, even though the RTX 20xx series does not support native BF16 compute (only storage); this may indicate emulation or a misleading runtime check.

In ComfyUI, --bf16-unet has no effect:
Model weight dtype: torch.bfloat16, manual cast: torch.float32

Observation:
The issue persists across differently quantized models (GGUF, FP8, SDNQ), suggesting the root cause is likely not quantization, but rather FP16 handling in the UNet or scheduler (e.g., overflow/underflow leading to NaNs in latents).

I hope something can be done here to fix the issue. Looking forward to your response.

**Comment by Disty0** [comment]
If you are on Linux or can use Triton, you can try SDNQ with INT8 matmul. Nvidia T4 runs 3x faster with INT8 compared to FP32 and 4x faster compared to BF16.

Here is the basic notebook on T4 which also doesn't support BF16 but supports INT8 just like RTX 2000 series: https://colab.research.google.com/drive/1A-4KX1S3PsUumuUD1SX-mO9IbuNyGlBA?usp=sharing

Key difference is enabling quantized matmul and setting torch_dtype to torch.float32 since T4 and RTX 2000 doesn't support BF16 and using BF16 breaks Triton on these GPUs.

**Comment by Mitrandin** [comment]
> If you are on Linux or can use Triton, you can try SDNQ with INT8 matmul. Nvidia T4 runs 3x faster with INT8 compared to FP32 and 4x faster compared to BF16.
> 
> Here is the basic notebook on T4 which also doesn't support BF16 but supports INT8 just like RTX 2000 series: https://colab.research.google.com/drive/1A-4KX1S3PsUumuUD1SX-mO9IbuNyGlBA?usp=sharing
> 
> Key difference is enabling quantized matmul and setting torch_dtype to torch.float32 since T4 and RTX 2000 doesn't support BF16 and using BF16 breaks Triton on these GPUs.

Thank you for helpfull response. Setting torch.float32 and use_quantized_matmul wtih triton installed indeed increase the speed to the same level as fp16 inference without breaking the image.
pipe = diffusers.ZImagePipeline.from_pretrained("Disty0/Z-Image-Turbo-SDNQ-uint4-svd-r32", torch_dtype=torch.float32)
pipe.transformer = apply_sdnq_options_to_model(pipe.transformer, use_quantized_matmul=True)
pipe.text_encoder = apply_sdnq_options_to_model(pipe.text_encoder, use_quantized_matmul=True)

Unfortunately, the same fix doesn't apply to comfyui. With "--fast fp8_matrix_mult --fp32-unet" flags no speed increase. May be it's different matmul algorithms or comfyui impementation incompatible with Z-image or my gpu.

**Comment by Ph0rk0z** [comment]
Run all your prompts from a python script is not an answer. I think everyone's inference code is busted.

**Comment by Disty0** [comment]
Same thing can be done in SD.Next UI (z-image support is in dev branch rn) by setting compute dtype to fp32 (do this only for gpus that doesn't support bf16) and enabling sdnq quantization and use quantized matmul in quantization settings if you don't want to use a python script.

**Comment by Ph0rk0z** [comment]
I did and I get about 8.9it/s vs comfyui 5it/s. I can't compile in next on turning either. Leaving it in "auto" was actually faster.

To clarify. FP32/quantized is 12it/s and "auto" bf16/quantized is 8.9it/s
